### PR TITLE
Add fixtures

### PR DESCRIPTION
**Pull request by irenedea**
Adds fixtures for models and boiler plate testing code.

TODO:

### CODE REVIEW

File: tests/fixtures/autouse.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+import gc
+
+import pytest
+import torch
+from composer.utils import dist, get_device, reproducibility
+
+
+@pytest.fixture(autouse=True)
```
Comment (irenedea) [inline tests/fixtures/autouse.py:11]: nit: either use naming pytest.fixture or just fixture across all fixture files

**Comment by irenedea** [comment]
hmm having trouble reproducing the CI code quality error locally, and the types it's complaining about seem to match correctly 🤷‍♀️ will take a closer look

File: tests/fixtures/data.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+from llmfoundry.data.finetuning.dataloader import build_finetuning_dataloader
+from tests.data_utils import make_tiny_ft_dataset
+
+
+@fixture
+def tiny_ft_dataset_path(tmp_path: Path, dataset_size: int = 4) -> str:
+    """Creates a tiny dataset and returns the path."""
+    tiny_dataset_path = os.path.join(tmp_path, 'test-ift-data-small')
+    os.mkdir(tiny_dataset_path)
+    tiny_dataset_file = os.path.join(tiny_dataset_path, 'train.jsonl')
```
Comment (aspfohl) [inline tests/fixtures/data.py:21]: Sort of a nit, but paths shouldn't require os at all
```suggestion
    tiny_dataset_path = tmp_path / 'test-ift-data-small'
    tiny_dataset_path.mkdir()
    tiny_dataset_file = tiny_dataset_path / 'train.jsonl'
```

File: tests/fixtures/autouse.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    """Clear memory between GPU tests."""
+    marker = request.node.get_closest_marker('gpu')
+    if marker is not None and torch.cuda.is_available():
+        torch.cuda.empty_cache()
+        gc.collect()  # Only gc on GPU tests as it 2x slows down CPU tests
+
+
+@pytest.fixture(autouse=True)
+def seed_all():
+    """Sets the seed for reproducibility."""
```
Comment (aspfohl) [inline tests/fixtures/autouse.py:33]: Nice, could all of the other seed calls get removed since this is autoused?

**Review by aspfohl — APPROVED** [review]
🌮

File: tests/fixtures/autouse.py (hunk 3 above)
Comment (irenedea) [reply tests/fixtures/autouse.py:33]: Yeah!! I'll do a pass to remove them

Comment (irenedea) [reply tests/fixtures/autouse.py:38]: [8848934](https://github.com/mosaicml/llm-foundry/pull/673/commits/8848934a3fbc91b581074af79ab45ff9d3f45859)

File: tests/fixtures/models.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    return model
+
+
+@fixture
+def mpt_tokenizer():
+    return build_tokenizer('EleutherAI/gpt-neox-20b', {})
+
+
+@fixture
+def build_mpt(
```
Comment (dakinggg) [inline tests/fixtures/models.py:30]:
```suggestion
def build_tiny_mpt(
```

File: tests/fixtures/models.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+from pytest import fixture
+from transformers import PreTrainedTokenizerBase
+
+from llmfoundry.models.hf.hf_causal_lm import ComposerHFCausalLM
+from llmfoundry.models.model_registry import COMPOSER_MODEL_REGISTRY
+from llmfoundry.models.mpt.modeling_mpt import ComposerMPTCausalLM
+from llmfoundry.utils.builders import build_tokenizer
+
+
+def _build_model(config: DictConfig, tokenizer: PreTrainedTokenizerBase,
```
Comment (dakinggg) [inline tests/fixtures/models.py:17]: I'm not sure the move to device should be contained in this fixture. Code that gets the model from the constructor should not actually assume that the model is moved to the right device. For testing, we often do move it, but I'm wondering if doing it automatically might hide some tricky bugs

File: tests/fixtures/models.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        config.update(kwargs)
+        model = _build_model(config, mpt_tokenizer, device)
+        assert isinstance(model, ComposerMPTCausalLM)
+        return model
+
+    return build
+
+
+@fixture
+def build_hf_mpt(
```
Comment (dakinggg) [inline tests/fixtures/models.py:51]:
```suggestion
def build_tiny_hf_mpt(
```

File: tests/fixtures/data.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+from pytest import fixture
+from torch.utils.data import DataLoader
+from transformers import PreTrainedTokenizerBase
+
+from llmfoundry.data.finetuning.dataloader import build_finetuning_dataloader
+from tests.data_utils import make_tiny_ft_dataset
+
+
+@fixture
+def tiny_ft_dataset_path(tmp_path: Path, dataset_size: int = 4) -> Path:
```
Comment (dakinggg) [inline tests/fixtures/data.py:16]: can we make this fixture clean up after itself?

File: tests/test_hf_mpt_gen.py [hunk 8]
```diff
@@ -1,68 +1,34 @@
...
-                        Any] = om.to_container(test_cfg.tokenizer,
-                                               resolve=True)  # type: ignore
-    tokenizer_name = tokenizer_cfg['name']
-    tokenizer_kwargs = tokenizer_cfg.get('kwargs', {})
-    tokenizer = build_tokenizer(tokenizer_name, tokenizer_kwargs)
-
-    # build model
-    model = COMPOSER_MODEL_REGISTRY[test_cfg.model.name](test_cfg.model,
-                                                         tokenizer)
-    test_cfg.n_params = sum(p.numel() for p in model.parameters())
```
Comment (dakinggg) [inline tests/test_hf_mpt_gen.py:65]: any idea why this line was here?

**Review by dakinggg — APPROVED** [review]
Just nits, definitely a step forward. Thanks for doing this!

File: tests/test_hf_mpt_gen.py (hunk 8 above)
Comment (irenedea) [reply tests/test_hf_mpt_gen.py:65]: yeah, not sure. I just removed because it was unused.

File: tests/fixtures/data.py (hunk 7 above)
Comment (irenedea) [reply tests/fixtures/data.py:16]: do you mean remove the dataset after the test is done? If so, I think the tmp_path fixture we are using will do that automatically

Comment (dakinggg) [reply tests/fixtures/data.py:16]: oh yeah, good point. Can you go remove the place where I explicitly `rmtree` the dataset?

Comment (irenedea) [reply tests/fixtures/data.py:16]: yup, can do that! also we should add a check to only create the dataset on rank 0, right?

Comment (irenedea) [reply tests/fixtures/data.py:18]: added logic below to only create the dataset on rank 0 if worldsize != 1

File: tests/fixtures/models.py (hunk 5 above)
Comment (irenedea) [reply tests/fixtures/models.py:17]: ah gotcha, let's stay on the safe side and move it out then.
